core/utils/kagome_db_editor.cpp

```diff
@@ -112,61 +157,101 @@ int main(int argc, char *argv[]) {
+                       common::Buffer(encoded_res.value()));
+    }
 
-    auto hash = RootHash::fromHexWithPrefix(argv[STATE_HASH]).value();
+    RootHash hash;
+    if (argc == 2) {
+      auto block_number = header.number + 1;
+      for (;;) {
+        auto header_res = block_storage->getBlockHeader(block_number++);
+        if (not header_res.has_value()) {
+          break;
+        }
+        hash = header_res.value().state_root;
+      }
+      log->trace("Autodetected best block number is #{}, state root is 0x{}",
+                 block_number - 1,
+                 hash.toHex());
+    } else {
+      hash = RootHash::fromHexWithPrefix(argv[STATE_HASH]).value();
+    }
 
     auto trie =
         TrieStorageImpl::createFromStorage(
-            hash,
             injector.template create<sptr<Codec>>(),
             injector.template create<sptr<TrieSerializer>>(),
             injector
                 .template create<sptr<storage::changes_trie::ChangesTracker>>())
             .value();
 
     if (COMPACT == cmd) {
-      auto batch = trie->getPersistentBatch().value();
-      auto cursor = batch->trieCursor();
-      auto res = cursor->next();
-      int count = 0;
-      {
-        TicToc t1("Process state.", log);
-        while (cursor->key().has_value()) {
-          count++;
-          res = cursor->next();
-        }
-      }
-      log->trace("{} keys were processed at the state.", ++count);
+      auto batch = persistent_batch(trie, hash);
+      auto finalized_batch = persistent_batch(trie, finalized_hash);
 
       auto db_cursor = storage->cursor();
       auto db_batch = storage->batch();
-      auto res2 = db_cursor->seek(prefix);
-      count = 0;
+      auto res = db_cursor->seek(prefix);
+      int count = 0;
       {
         TicToc t2("Process DB.", log);
         while (db_cursor->isValid() && db_cursor->key().has_value()
                && db_cursor->key().value()[0] == prefix[0]) {
-          res = db_batch->remove(db_cursor->key().value());
+          auto res2 = db_batch->remove(db_cursor->key().value());
           count++;
           if (not(count % 10000000)) {
             log->trace("{} keys were processed at the db.", count);
-            res = db_batch->commit();
+            res2 = db_batch->commit();
             dynamic_cast<storage::LevelDB *>(storage.get())
                 ->compact(prefix, db_cursor->key().value());
             db_cursor = storage->cursor();
             db_batch = storage->batch();
-            res2 = db_cursor->seek(prefix);
+            res = db_cursor->seek(prefix);
           }
-          res = db_cursor->next();
+          res2 = db_cursor->next();
         }
-        res = db_batch->commit();
+        [[maybe_unused]] auto res2 = db_batch->commit();
       }
       log->trace("{} keys were processed at the db.", ++count);
 
       {
         TicToc t3("Commit state.", log);
-        auto res3 = batch->commit();
+        auto res3 = finalized_batch->commit();
```

Comment:
Why cases of failure is not processing?

Reply:
It is left to the backtrace lib to leave meaningful error on failure. Any non-result causes utility failure.

Reply:
Well, it always be segfault at some position:) I'll think about the error handling manner.

Reply:
I added checks "if no value throw the error"